minplascalc/species.py

```diff
@@ -0,0 +1,306 @@
+
+    def partitionfunction_internal(self, T, dE):
+        raise NotImplementedError
+
+    def internal_energy(self, T, dE):
+        raise NotImplementedError
+
+
+class Species(BaseSpecies):
+    def __init__(self, name, stoichiometry, molarmass, chargenumber):
+        """Base class for heavy particles. Either single monatomic or diatomic 
+        chemical species in the plasma, eg O2 or Si+
+
+        Parameters
+        ----------
+        name : string
+            A unique identifier for the species.
+        stoichiometry : dictionary
+            Dictionary describing the elemental stoichiometry of the species.
+        molarmass : float
+            Molar mass of the species in kg/mol.
+        chargenumber : int
+            Charge on the species (in integer units of the fundamental charge).
+        """
+        self.name = name
+        self.stoichiometry = deepcopy(stoichiometry)
+        self.molarmass = molarmass
+        self.chargenumber = chargenumber
+        if self.chargenumber < 0:
+            raise ValueError('Negatively charged ions not implemented.')
+
+
+class Monatomic(Species):
+    def __init__(self, name, stoichiometry, molarmass, chargenumber, 
+                 ionisationenergy, energylevels, sources):
+        """Class for monatomic plasma species (single atoms and ions).
+    
+        Parameters
+        ----------
+        name : string
+            A unique identifier for the species.
+        stoichiometry : dictionary
+            Dictionary describing the elemental stoichiometry of the species 
+            (e.g. {'O': 1} for O or O+).
+        molarmass : float
+            Molar mass of the species in kg/mol.
+        chargenumber : int
+            Charge on the species (in integer units of the fundamental charge).
+        ionisationenergy : float
+            Ionisation energy of the species in J.
+        energylevels : list of length-2 lists
+            Atomic energy level data - each entry in the list contains a pair of
+            values giving the level's quantum number and its energy
+            respectively, with energy in J.
+        sources : list of str
+            Each entry represents a reference from which the data was
+            obtained.
+        """
+        super().__init__(name, stoichiometry, molarmass, chargenumber)
+        
+        self.ionisationenergy = ionisationenergy
+        self.energylevels = deepcopy(energylevels)
+        self.sources = deepcopy(sources)
+
+    def __repr__(self):
+        return ('MonatomicSpecies(name=' + self.name + ','
+                + 'stoichiometry=' + str(self.stoichiometry) + ','
+                + 'molarmass=' + str(self.molarmass) + ','
+                + 'chargenumber=' + str(self.chargenumber) + ','
+                + 'ionisationenergy=' + str(self.ionisationenergy) + ','
+                + 'energylevels=' + str(self.energylevels) + ','
+                + 'sources=' + str(self.sources) + ')')
+
+    def __str__(self):
```

Comment:
Looking at these it seems like we can cut out a ton of boilerplate by using [dataclass](https://docs.python.org/3/library/dataclasses.html). But at least look into using f strings instead of these ugly string conversions.

Reply:
Dataclasses look very nifty indeed. It would likely need a fairly substantial refactor, but probably worth it if it makes the code simpler and cleaner. For now I'll implement f-strings, that's a good alternative, thanks.